corehq/apps/userreports/models.py

```diff
@@ -146,6 +146,29 @@ class DataSourceActionLog(models.Model):
     skip_destructive = models.BooleanField(default=False)
 
 
+class DataSourceRowTransactionLog(models.Model):
+    CREATE = "create"
+    UPDATE = "update"
+    DELETE = "delete"
+
+    domain = models.CharField(max_length=126, null=False, db_index=True)
+    date_created = models.DateTimeField(auto_now_add=True)
+    action = models.CharField(max_length=32, choices=(
+        (CREATE, CREATE),
+        (UPDATE, UPDATE),
+        (DELETE, DELETE),
+    ), null=False, blank=False)
+    data_source_id = models.UUIDField(null=False, blank=False)
+    row_id = models.UUIDField(unique=True, null=False, blank=False, db_index=True)  # The Form or Case ID
```

Comment:
I made this field unique since we're only retaining one record of each row.

Reply:
(Sadly) both `XFormInstance.form_id` and `CommCareCase.case_id` are `CharFields`. We won't know whether to fetch cases and forms by `str(row_id)` or by `row_id.hex` :/

So this should probably be a `CharField` too.